ledger-service/http-json/src/main/scala/com/digitalasset/http/json/ResponseFormats.scala

```diff
@@ -4,74 +4,79 @@
+import akka.http.scaladsl.model.StatusCode
+import akka.http.scaladsl.model.StatusCodes.{InternalServerError, OK}
+import akka.stream.scaladsl.{Concat, Source}
 import akka.util.ByteString
-import com.daml.fetchcontracts.util.AkkaStreamsDoobie
 import scalaz.syntax.show._
-import scalaz.{Show, \/}
+import scalaz.{Show, -\/, \/, \/-}
 import spray.json.DefaultJsonProtocol._
 import spray.json._
 
 private[http] object ResponseFormats {
   def errorsJsObject(status: StatusCode, es: String*): JsObject = {
     val errors = es.toJson
-    JsObject(statusField(status), ("errors", errors))
+    JsObject(("status", JsNumber(status.intValue)), ("errors", errors))
   }
 
   def resultJsObject[A: JsonWriter](a: A): JsObject = {
     resultJsObject(a.toJson)
   }
 
   def resultJsObject(a: JsValue): JsObject = {
-    JsObject(statusField(StatusCodes.OK), ("result", a))
+    JsObject(("status", JsNumber(OK.intValue)), ("result", a))
   }
 
   def resultJsObject[E: Show](
       jsVals: Source[E \/ JsValue, NotUsed],
       warnings: Option[JsValue],
   ): Source[ByteString, NotUsed] = {
-
-    val graph = GraphDSL.create() { implicit b =>
-      import GraphDSL.Implicits._
-
-      val partition: FanOutShape2[E \/ JsValue, E, JsValue] = b add AkkaStreamsDoobie.partition
-      val concat: UniformFanInShape[ByteString, ByteString] = b add Concat(3)
-
-      // first produce optional warnings and result element
-      warnings match {
-        case Some(x) =>
-          Source.single(ByteString(s"""{"warnings":${x.compactPrint},"result":[""")) ~> concat.in(0)
-        case None =>
-          Source.single(ByteString("""{"result":[""")) ~> concat.in(0)
+    jsVals
+      // Collapse the stream of `E \/ JsValue` into a single pair of errors and results,
+      // only one of which may be non-empty.
+      .fold((Vector.empty[E], Vector.empty[JsValue])) {
+        case ((errors, results), \/-(r)) if errors.isEmpty => (Vector.empty, results :+ r)
+        case ((errors, _), \/-(_)) => (errors, Vector.empty)
+        case ((errors, _), -\/(e)) => (errors :+ e, Vector.empty)
       }
+      // Convert that into a stream containing the appropriate JSON response object
+      .flatMapConcat {
+        case (errors, results) => {
+          val comma = single(",")
+          val (payload, status) = {
+            val (name, vals, statusCode): (String, Iterator[JsValue], StatusCode) =
+              if (errors.nonEmpty)
+                ("errors", errors.iterator.map(e => JsString(e.shows)), InternalServerError)
+              else
+                ("result", results.iterator, OK)
+
+            (arrayField(name, vals), scalarField("status", JsNumber(statusCode.intValue)))
+          }
+          Source.combine(
+            single("{"),
+            warnings.fold(Source.empty[ByteString])(scalarField("warnings", _) ++ comma),
+            payload,
+            comma,
+            status,
+            single("}"),
+          )(Concat(_))
```

Comment:
Nit: The current code is definitely more targeted than the previous version, but could maybe go even further in my opinion. No need really to use the Akka Streams machinery to concatenate a string. One could just call `.map` instead of `flatMapConcat` and construct the ByteString the conventional way, something like this (untested):

```
  def resultJsObject[E: Show](
      jsVals: Source[E \/ JsValue, NotUsed],
      warnings: Option[JsValue],
  ): Source[ByteString, NotUsed] = {
    jsVals
      // Collapse the stream of `E \/ JsValue` into a single pair of errors and results,
      // only one of which may be non-empty.
      .fold((Vector.empty[E], Vector.empty[JsValue])) {
        case ((errors, results), \/-(r)) if errors.isEmpty => (Vector.empty, results :+ r)
        case ((errors, _), \/-(_)) => (errors, Vector.empty)
        case ((errors, _), -\/(e)) => (errors :+ e, Vector.empty)
      }
      // Convert that into a stream containing the appropriate JSON response object
      .map { case (errors, results) =>
        val (payload, status) = {
          val (name, vals, statusCode): (String, Seq[JsValue], StatusCode) =
            if (errors.nonEmpty)
              ("errors", errors.map(e => JsString(e.shows)), InternalServerError)
            else
              ("result", results, OK)
          (arrayField(name, vals), scalarField("status", JsNumber(statusCode.intValue)))
        }
        ByteString(
          (warnings.map(v => scalarField("warnings", v)).toList ::: payload :: status :: Nil)
            .mkString("{", ",", "}")
        )
      }
  }

  private def scalarField(name: String, value: JsValue) = s""""${name}":${value.compactPrint}"""

  private def arrayField(name: String, items: Seq[JsValue]): String = {
    val csv = items.zipWithIndex.map { case (r, i) =>
      val str = r.compactPrint
      if (i == 0) str else "," + str
    }
    s""""${name}":[$csv]"""
  }
  ```

Reply:
@raphael-speyer-da pointed out that the the ByteString should still be in small chunks to not exceed the HTTP chunk size limit, and that the `.iterator` approach is good to make the collections lazy. One could adapt above code accordingly, but at that point it's probably best just to leave things as they are...

Reply:
I think adapting the code accordingly would bring it to approximately what's in this PR anyway.